File: presentation/src/main/java/by/grodno/vasili/githubsimpleapp/feature/users/UsersDependenciesModule.java
package by.grodno.vasili.githubsimpleapp.feature.users;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import by.grodno.vasili.data.datasource.retrofit.RetrofitUserDatasource;
import by.grodno.vasili.data.repository.UserDataRepository;
import by.grodno.vasili.domain.interactor.GetUsersUseCase;
import by.grodno.vasili.githubsimpleapp.thread.IOThread;
import by.grodno.vasili.githubsimpleapp.thread.UIThread;

/**
 * Dependency factory for {@link UsersActivity}
 */
class UsersDependenciesModule {
    private static final DiffUtil.ItemCallback<UserItem> DIFF_CALLBACK = new DiffUtil.ItemCallback<UserItem>() {
        @Override
        public boolean areItemsTheSame(@NonNull UserItem oldItem, @NonNull UserItem newItem) {
            return oldItem.id.equals(newItem.id);
        }

        @Override
        public boolean areContentsTheSame(@NonNull UserItem oldItem, @NonNull UserItem newItem) {
            return oldItem.equals(newItem);
        }
    };
    private final UsersViewModelFactory factory;
    private final UsersAdapter adapter;

    UsersDependenciesModule() {
        GetUsersUseCase getUsersUseCase = new GetUsersUseCase(
                new IOThread(),
                new UIThread(),
                new UserDataRepository(new RetrofitUserDatasource())
        );
        UserItemDatasourceFactory datasourceFactory = new UserItemDatasourceFactory(getUsersUseCase);
        factory = new UsersViewModelFactory(getUsersUseCase, datasourceFactory);
        adapter = new UsersAdapter(DIFF_CALLBACK);
    }

    UsersViewModelFactory getFactory() {
        return factory;
    }

    UsersAdapter getAdapter() {
        return adapter;
    }
}

File: presentation/src/main/java/by/grodno/vasili/githubsimpleapp/feature/users/UsersDependenciesModule.kt
package by.grodno.vasili.githubsimpleapp.feature.users

import androidx.recyclerview.widget.DiffUtil
import by.grodno.vasili.data.datasource.retrofit.RetrofitUserDatasource
import by.grodno.vasili.data.repository.UserDataRepository
import by.grodno.vasili.domain.interactor.GetUsersUseCase
import by.grodno.vasili.githubsimpleapp.BuildConfig
import kotlinx.coroutines.CoroutineScope

/**
 * Dependency factory for [UsersActivity]
 */
internal class UsersDependenciesModule(scope: CoroutineScope) {
    val factory: UsersViewModelFactory
    val adapter: UsersAdapter

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<UserItem> = object : DiffUtil.ItemCallback<UserItem>() {
            override fun areItemsTheSame(oldItem: UserItem, newItem: UserItem): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: UserItem, newItem: UserItem): Boolean {
                return oldItem == newItem
            }
        }
    }

    init {
        val getUsersUseCase = GetUsersUseCase(UserDataRepository(RetrofitUserDatasource(BuildConfig.GITHUB_TOKEN)))
        val datasourceFactory = UserItemDatasourceFactory(getUsersUseCase, scope)
        factory = UsersViewModelFactory(getUsersUseCase, datasourceFactory)
        adapter = UsersAdapter(DIFF_CALLBACK)
    }
}